File: Core/PermissionDatabaseCatalogue.cs
using KolibSoft.AuthStore.Core.Models;
using KolibSoft.Catalogue.Core;
using Microsoft.EntityFrameworkCore;

namespace KolibSoft.AuthStore.Core;

public class PermissionDatabaseCatalogue : DatabaseCatalogue<PermissionModel, PermissionFilters>
{

    public DbSet<CredentialPermissionModel> CredentialPermissions { get; }

    protected override IQueryable<PermissionModel> QueryItems(IQueryable<PermissionModel> items, PermissionFilters? filters = default)
    {
        if (filters?.Clean ?? true) items = items.Where(x => x.Active);
        if (filters?.Hint != null) items = items.Where(x => EF.Functions.Like(x.Code, $"%{filters.Hint}%"));
        items = items.OrderBy(x => x.Code).OrderByDescending(x => x.Active);
        return items;
    }

    protected override bool ValidateInsert(PermissionModel item)
    {
        if (DbSet.Any(x => x.Code == item.Code))
        {
            Errors?.Add(AuthStoreStatics.RepeatedCode);
            return false;
        }
        return true;
    }

    protected override bool ValidateUpdate(PermissionModel item)
    {
        if (DbSet.Any(x => x.Code == item.Code && x.Id != item.Id))
        {
            Errors?.Add(AuthStoreStatics.RepeatedCode);
            return false;
        }
        return true;
    }

    protected override bool ValidateDelete(PermissionModel item)
    {
        if (CredentialPermissions.Any(x => x.PermissionId == item.Id))
        {
            Errors?.Add(CatalogueStatics.UsedItem);
            return false;
        }
        return true;
    }

    public PermissionDatabaseCatalogue(DbContext dbContext) : base(dbContext)
    {
        CredentialPermissions = dbContext.Set<CredentialPermissionModel>();
    }

}

File: docs/lib/catalogues/permission_database_catalogue.js
import { AuthStoreStatics } from "../authstore_statics.js";
import { CredentialPermissionModel } from "../models/credential_permission_model.js";
import { PermissionModel } from "../models/permission_model.js";
import { CatalogueStatics, DatabaseCatalogue, DbContext, DbSet } from "../modules/catalogue.js";
import { PermissionFilters } from "../filters/permission_filters.js";

class PermissionDatabaseCatalogue extends DatabaseCatalogue {

    /** @type {DbSet<CredentialPermissionModel>} */
    #credentialPermissions;

    /** @returns {DbSet<CredentialPermissionModel>} */
    get credentialPermissions() { return this.#credentialPermissions; }

    /**
     * @param {PermissionModel[]} items 
     * @param {PermissionFilters} filters 
     * @returns {Promise<PermissionModel[]>}
     */
    async queryItems(items, filters = null) {
        if (filters?.clean) items = items.filter(x => x.active);
        if (filters?.hint) items = items.filter(x => x.code.includes(filters.hint));
        items = items.sort((a, b) => a.code.localeCompare(b.code)).sort((a, b) => b.active - a.active);
        return items;
    }

    /**
     * @param {PermissionModel} item 
     * @returns {boolean}
     */
    async vlidateInsert(item) {
        if ((await this.dbSet.filter(x => x.code == item.code)).length > 0) {
            this.errors.push(AuthStoreStatics.RepeatedIdentity);
            return false;
        }
        return true;
    }

    /**
     * @param {PermissionModel} item 
     * @returns {boolean}
     */
    async validateUpdate(item) {
        if ((await this.dbSet.filter(x => x.code == item.code && x.id != item.id)).length > 0) {
            this.errors.push(AuthStoreStatics.RepeatedIdentity);
            return false;
        }
        return true;
    }

    /**
     * @param {PermissionModel} item 
     * @returns {boolean}
     */
    async validateDelete(item) {
        if ((await this.#credentialPermissions.filter(x => x.permissionId == item.id)).length > 0) {
            this.errors.push(CatalogueStatics.UsedItem);
            return false;
        }
        return true;
    }

    /**
     * @param {(json: {} => PermissionModel)} creator 
     * @param {DbContext} dbContext 
     * @param {string} name 
     */
    constructor(creator, dbContext, name) {
        super(creator, dbContext, name);
        this.#credentialPermissions = dbContext.set("credential_permission");
    }

}

export {
    PermissionDatabaseCatalogue
}